File: internal/check/ripper.go
// Package check provides ripper detection for CD-rip log files.
package check

import (
	"errors"
	"regexp"
	"strings"
)

const (
	Unknown    = "unknown"
	Whipper    = "whipper"
	XLD        = "XLD"
	EAC        = "EAC"
	DBpoweramp = "dBpoweramp"
)

// ErrUnknownRipper is returned when the log cannot be attributed to a known ripper.
var ErrUnknownRipper = errors.New("could not determine ripper")

var dbpowerampRe = regexp.MustCompile(`(?im)^dBpoweramp Release`)

// GetRipper inspects log text and returns one of the ripper constants.
func GetRipper(log string) (string, error) {
	if strings.Contains(log, "Log created by: whipper") {
		return Whipper, nil
	}
	if strings.Contains(log, "X Lossless Decoder version") {
		return XLD, nil
	}
	if strings.Contains(log, "Exact Audio Copy") {
		return EAC, nil
	}
	if dbpowerampRe.MatchString(log) {
		return DBpoweramp, nil
	}
	// Fallback: check first line for "EAC"
	if nl := strings.Index(log, "\n"); nl != -1 {
		if strings.Contains(log[:nl], "EAC") {
			return EAC, nil
		}
	}
	return Unknown, ErrUnknownRipper
}

File: src/Check/Ripper.php
<?php

declare(strict_types=1);

namespace OrpheusNET\Logchecker\Check;

use OrpheusNET\Logchecker\Exception\UnknownRipperException;

class Ripper
{
    public const UNKNOWN = 'unknown';
    public const WHIPPER = 'whipper';
    public const XLD = 'XLD';
    public const EAC = 'EAC';
    public const DBPOWERAMP = 'dBpoweramp';

    public static function getRipper(string $log): string
    {
        if (strpos($log, "Log created by: whipper") !== false) {
            return Ripper::WHIPPER;
        } elseif (strpos($log, "X Lossless Decoder version") !== false) {
            return Ripper::XLD;
        } elseif (strpos($log, "Exact Audio Copy") !== false) {
            return Ripper::EAC;
        } elseif (preg_match('/^dBpoweramp Release/im', $log)) {
            return Ripper::DBPOWERAMP;
        } else {
            $firstLine = strstr($log, "\n", true);
            if ($firstLine !== false && strpos($firstLine, "EAC") !== false) {
                return Ripper::EAC;
            } else {
                throw new UnknownRipperException("Could not determine ripper");
            }
        }
    }
}
